### PR TITLE
Add support for Learning Mode

### LINKED ISSUE #46: Learning Mode - Style tweaks for all Learning Mode templates
Update the Learning Mode templates so that they match the [design](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20020082%3A4588). Note that there's a [separate issue](https://github.com/Automattic/sensei-theme/issues/47) to make more style tweaks to individual templates. As part of this, the theme will need to opt out of opinionated Learning Mode styles so that theme styles are applied instead.

Following are the common elements that need to be updated across all templates. It's possible that not all of these UI elements will be able to be targeted in `theme.json`. The sidebar in particular was mentioned as something that may be tricky to customize.

Also, once https://github.com/Automattic/sensei-theme/issues/48 is implemented, some of these styling tweaks may no longer be necessary.

## Header
_Course title_ - Font family, font weight, font size, uppercase
_Exit course link_ - Font family, font size, underline

## Sidebar (Course Outline)
_Module title_ - Uppercase, font family, font weight, font size
_Completed status indicator (checkmark)_ - No fill, green checkmark
_Lesson title_ - Font family, font weight, font size
_Quiz label_ - Font family, font size
_Collapsed lesson content_ - Uppercase, font family, font size

## Buttons
Font family, font weight, font size, border radius, uppercase

---
_Note that any inconsistencies you may notice in the screenshots that are not already documented above are likely a result of designs that have not been updated since the learning mode template project was completed._

## Default Template
Screen Shot 2022-10-20 at 9 47 37 AM

## Modern Template
Screen Shot 2022-10-20 at 9 48 26 AM

### Video
Screen Shot 2022-10-20 at 9 49 12 AM

### No Video
Screen Shot 2022-10-20 at 10 14 27 AM

## Large Video Template
Screen Shot 2022-10-20 at 9 50 25 AM

#### Scenario: Default template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Default Learning Mode template is activated
**When** I view a lesson
**Then** the page looks similar to the Default Template screenshot.

#### Scenario: Modern template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Modern Learning Mode template is activated
**When** I view a lesson
**Then** the page looks similar to the Modern Template screenshot.

#### Scenario: Video template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Video Learning Mode template is activated
**When** I add the Featured Video block to a lesson
**And** I view that lesson on the frontend
**Then** the page looks similar to the Video Template - Video screenshot.

#### Scenario: Video template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Video Learning Mode template is activated
**When** I view a lesson that doesn't contain the Featured Video block
**Then** the page looks similar to the Video Template - No Video screenshot.

#### Scenario: Large Video template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Large Video Learning Mode template is activated
**When** I add the Featured Video block to a lesson
**And** I view that lesson on the frontend
**Then** the page looks similar to the Large Video Template screenshot.

### LINKED ISSUE #47: Learning Mode - Style tweaks for individual Learning Mode templates
Dependent on #48.

Note that there's a [separate issue](https://github.com/Automattic/sensei-theme/issues/46) for adjusting common styles across all Learning Mode templates.

Once https://github.com/Automattic/sensei-theme/issues/48 is implemented, some of these styling tweaks may no longer be necessary.

## Video Template

Update the Video Learning Mode template to match the [design](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20020082%3A4588).

In particular, the course progress bar text needs to be in uppercase, and the font family and font size updated. The progress bar itself also needs to be taller and the border radius adjusted.

Screen Shot 2022-10-20 at 1 57 39 PM

## Default & Large Video Templates

Update the Default and Large Video templates to match the [design](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20020083%3A5966).

In particular, the font family, font size and color of the "2 of 10 lessons complete (25%)" text in the header needs to be updated. Also, the background color of the progress bar needs to be tweaked (the fill color is correct).

Screen Shot 2022-10-20 at 2 06 00 PM

## Modern & Video & Large Video Templates

Update the [Modern](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20021361%3A11212), [Video](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20020082%3A4588) and [Large Video](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20020093%3A1715) templates to match the design.

In particular, the border radius around the video + course outline container in the Video template, and the border radius around the course outline container in the Modern and Large Video templates need to be adjusted.

Screen Shot 2022-10-20 at 2 12 01 PM

Screen Shot 2022-10-20 at 2 14 06 PM

#### Scenario: Default template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Default Learning Mode template is activated
**When** I view a lesson
**Then** the header text and progress bar background color look similar to the design.

#### Scenario: Modern template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Modern Learning Mode template is activated
**When** I view a lesson
**Then** the border radius around the course outline container looks similar to the design.

#### Scenario: Video template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Video Learning Mode template is activated
**When** I add the Featured Video block to a lesson
**And** I view that lesson on the frontend
**Then** the border radius around the video + course outline container looks similar to the design
**And** the course progress bar text is in uppercase and has the correct font family and size
**And** the progress bar has the correct height and border radius.

#### Scenario: Large Video template
**Given** The Sensei theme is activated
**And** Learning Mode is enabled for all courses
**And** the Large Video Learning Mode template is activated
**When** I add the Featured Video block to a lesson
**And** I view that lesson on the frontend
**Then** the header text and progress bar background color look similar to the design
**And** the border radius around the course outline container looks similar to the design.

### PR DESCRIPTION
**Pull request by gabrielcaires**
Resolves https://github.com/Automattic/course/issues/46 , https://github.com/Automattic/course/issues/47

## Changes proposed in this pull request

- Updated theme.json and styles to make free and paid Learning Mode templates match the design

## Testing instructions

- Checkout this brunch of the Sensei Plugin https://github.com/Automattic/sensei/pull/6068
- Have the latest trunk of Sensei Pro
- Make sure when you go to sensei settings, you get all 4 templates for LM
- Make sure to enable LM
- Create a course with several modules, containing several lessons each
- Make sure to add course videos in some of those lessons, also don't add any videos in some of them
- Add quizzes in some of the lessons so that the "Take Quiz" button appears at the bottom
- Now check the course as a student
- Make sure to switch to all templates and for each of the templates, check at least one lesson with a video and one without
- Make sure the designs look as expected, with acceptably close font size and spacings to the design and overall structure matches

## Screenshots

**Default**
Screen Shot 2022-11-19 at 4 21 50 AM

**Video**
Screen Shot 2022-11-19 at 4 20 55 AM

**Modern**
Screen Shot 2022-11-19 at 4 21 24 AM

**Large Video**
Screen Shot 2022-11-19 at 4 23 27 AM

### CODE REVIEW

File: assets/icons/completed.svg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<svg fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (yscik) [inline assets/icons/completed.svg]: Icon sizes are broken for this, needs a viewBox attribute (`viewBox="0 0 12 12"`). Note that we have been using a 24x24 authored size for all the other icons, might be better to re-export in that size. (Though not sure if there is any difference when setting the size via viewbox.)

image

File: learning-mode.css [hunk 2]
```diff
@@ -0,0 +1,282 @@
+:root {
+	--lm-header-height: 116px;
```
Comment (yscik) [inline learning-mode.css]: This height, together with the padding, doesn't seem right with templates outside Modern:

#### Default:
image

#### Video:
image

#### Large video:
image

#### Editor w/ Default:
image

File: assets/js/header.js [hunk 3]
```diff
@@ -5,4 +5,12 @@ const observer = new IntersectionObserver(
   { threshold: [1] }
 );
 
-observer.observe( headerElement );
\ No newline at end of file
+if ( headerElement ) {
+  observer.observe( headerElement );
+}
+
+const isFullWidthLearningModeTemplate = document.querySelector( '.sensei-course-theme__header .sensei-course-theme-course-progress-bar' );
```
Comment (yscik) [inline assets/js/header.js:12]: Should we add a class to the body in the Sensei plugin indicating which template is being used, to make styling them easier? Though that won't work for user-built templates. 

For this specific case, we wanted to do container-wide headers instead of full-width in the Modern/Video templates, but had to scope it out because it was tricky to make it work across themes and versions. But it should be worth doing if it also helps this theme.

File: styles/gold.json [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+			"fontFamilies": [
+				{
+					"fontFamily": "Arvo",
+					"name": "Arvo",
+					"slug": "body",
+					"fontFace": [
+						{
+							"fontFamily": "Arvo",
+							"fontWeight": "400",
+							"src": [ "file:./assets/fonts/arvo/Arvo-Regular.ttf" ]
```
Comment (yscik) [inline styles/gold.json]: Not part of the PR, but worth converting these fonts to `.woff` format instead, which is supported in all browsers and can get 2-3x smaller file sizes.

File: learning-mode.css [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+	font-family: var(--wp--preset--font-family--heading);
+	font-size: var(--wp--preset--font-size--medium);
+	line-height: 100%;
+	letter-spacing: 0.01em;
+	text-transform: uppercase;
+	color: var(--wp--preset--color--foreground);
+	text-decoration: none;
+}
+
+.sensei-lms-course-navigation-lesson__extra {
```
Comment (yscik) [inline learning-mode.css:46]: The quiz link seems just a tiny bit misaligned:

image

File: learning-mode.css [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    font-family: var(--wp--preset--font-family--system);
+    line-height: 100%;
+    font-weight: normal;
+    opacity: 1;
+    letter-spacing: 0.02em;
+}
+
+.sensei-lms-course-navigation-module__summary {
+	font-family: var(--wp--preset--font-family--system);
+	font-size: 11px;
```
Comment (yscik) [inline learning-mode.css:57]: The summary text is a bit small:
image

Should be 16px uppercase if I'm looking at the right design, with less margin?
image

File: learning-mode.css [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+	--border-color:  var(--border-color)
+}
+
+.sensei-lms-course-navigation-lesson {
+	padding-top: 20px;
+	padding-bottom: 0;
+}
+
+.sensei-course-theme__sidebar {
+	--sidebar-width: 280px;
```
Comment (yscik) [inline learning-mode.css]: Is this the correct sizing? The designs suggest wider, around 350px. 

Note for the paddings that the sidebar content also has some extra spacing on the sides due to the [scrollbar-gutter](https://developer.mozilla.org/en-US/docs/Web/CSS/scrollbar-gutter) rule, to keep things stable and symmetric when the scrollbar is on/off.

File: learning-mode.css [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+}
+.sensei-course-theme-lesson-actions .sensei-course-theme__button.has-icon svg,
+.sensei-course-theme-lesson-actions .sensei-course-theme__link.has-icon svg {
+  width: 20px;
+  height: 20px;
+}
+
+.sensei-course-theme__main-content > div.wp-block-group:last-child,
+.sensei-course-theme-lesson-actions {
+  flex-direction: row-reverse;
```
Comment (yscik) [inline learning-mode.css]: I'm not sure CSS should change the order of the buttons. That brings up some accessibility issues (tab order won't change). The lesson actions block also makes sure the primary button is first. It also breaks alignment and pagination.

image

Original:

image

Was this change intentional in the design @Luchadores?

Design:

image

File: learning-mode.css [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+.sensei-lms-course-navigation-module:not(:first-of-type) {
+	margin-top: 18px;
+}
+
+.wp-block-sensei-button:where(:not(.is-style-link)) a:hover{
+	text-decoration: none;
+}
+
+/*
```
Comment (yscik) [inline learning-mode.css:77]: When a button is focused, it shrinks:

Screen Recording 2022-11-22 at 21 03 11

File: theme.json [hunk 10]
```diff
@@ -415,17 +464,17 @@
...
+						"top": "1.25rem",
+						"bottom": "1.25rem"
 					}
 				},
+				"typography": {
+					"lineHeight": "clamp(1.838rem, 1.763rem + 0.249vw, 1.95rem)",
+					"letterSpacing": "-0.01em"
+				},
 				"elements": {
 					"link": {
```
Comment (yscik) [inline theme.json:479]: Links in content don't have hover styles. 
Looks like it's also missing in a few other places - Exit Course block, and outside learning mode in the content, main menu and sidebar links.

**Review by yscik — COMMENTED** [review]
Nice work on this, looks pretty good! I mostly found styling issues:

**Comment by Luchadores** [comment]
Hey @yscik 👋 I noticed this GitHub issue and would like to link here the designs as well we have for Learning Mode, to have a look at spacing and all of that. 

Figma file [here](https://www.figma.com/file/nf4oV51c1JBxJKXshMe7cd/New-Sensei-Theme?node-id=20020040%3A18973&t=Rn8vL7yvHWq7M4hw-1).

image

**Comment by Imran92** [comment]
Hi @Luchadores 👋 I think @yscik was [pointing to the button orientation and ordering](https://github.com/Automattic/course/pull/68#discussion_r1029767538) and I also have the same question. Because though we can implement it using CSS, this is a template-level change

> I'm not sure CSS should change the order of the buttons. That brings up some accessibility issues (tab order won't change). The lesson actions block also makes sure the primary button is first. It also breaks alignment and pagination.

> Was this change intentional in the design @Luchadores?

**Comment by Luchadores** [comment]
👋 @Imran92 

Thanks for pointing that out. I believe the main CTA should always be on the right side and the secondary button on the left.
We can keep them as they are for now and see if and when we can update them in the future.

File: learning-mode.css (hunk 2 above)
Comment (Imran92) [reply learning-mode.css]: Hi @yscik 👋 Are you referring to the figma design here or the implementation? As in the design, the height is 116px
image

File: learning-mode.css (hunk 7 above)
Comment (Imran92) [reply learning-mode.css]: I tried with 350px first according to the design. Then it looked like this width is actually works as the content width of the sidebar, and the content width in the design for the sidebar is around 280-290px depending on the template. I've updated it to 290px [here](https://github.com/Automattic/course/pull/68/commits/6599db0963b20462304a2c69ff550e1dbac09e05) to get a bit more space

Screen Shot 2022-11-23 at 10 18 37 PM

File: learning-mode.css (hunk 8 above)
Comment (Imran92) [reply learning-mode.css]: Removed it [here](https://github.com/Automattic/course/pull/68/commits/03a6ad31f10d97ba2b223ef70e7307262f5ecfbe) after discussion

File: learning-mode.css (hunk 9 above)
Comment (Imran92) [reply learning-mode.css:77]: Good find! Fixed it [here](https://github.com/Automattic/course/pull/68/commits/21dea191eea4d725f2c80ef9e9aacb469f0c8e58)

File: learning-mode.css (hunk 5 above)
Comment (Imran92) [reply learning-mode.css:46]: Good catch, fixed [here](https://github.com/Automattic/course/pull/68/commits/a2d62d33d322d727cdc05171881bf6e774169ff8)

File: learning-mode.css (hunk 6 above)
Comment (Imran92) [reply learning-mode.css:57]: I've updated the margin [here](https://github.com/Automattic/course/pull/68/commits/687b0afca5b441300d4f0d1aa4e3f285738da1b8) and made it uppercase. But in the design, it is 11px I think but the line height is 16px
Screen Shot 2022-11-23 at 10 15 49 PM

File: assets/icons/completed.svg (hunk 1 above)
Comment (Imran92) [reply assets/icons/completed.svg]: The icon changes were [removed](https://github.com/Automattic/course/pull/68/commits/d59c65888192ce60f6a8d933429080855052e155) after some discussion, but learned something new here, thanks!

File: assets/js/header.js (hunk 3 above)
Comment (Imran92) [reply assets/js/header.js:12]: I was thinking the same, because found it really hard to differentiate among templates, as you said, it was tricky. To avoid editing templates, thought about inserting the name of the template as a class manipulating string using code where the template files are loaded. But then decided to use JS to not edit all templates in the core plugin and pro now as it was working for now.

But I think we should add those classes later and also remove some inline styles from the templates so that we don't have to struggle overriding those from themes and have to resort to using !important

File: styles/gold.json (hunk 4 above)
Comment (Imran92) [reply styles/gold.json]: Thanks for pointing this out! I'll create a ticket for this and change all the fonts under that

File: theme.json (hunk 10 above)
Comment (Imran92) [reply theme.json:479]: Good catch! Fixed the ones pertaining to learning mode [here](https://github.com/Automattic/course/pull/68/commits/62b93289217c50c3f55a2357b99489c40a205aad)

**Comment by Imran92** [comment]
Thanks a lot for taking such a close look @yscik ! I've updated the PR addressing your comments, can you kindly take a look again? Thanks!

**Review by Imran92 — APPROVED** [review]
